[14.0][IMP] l10n_br_fiscal, l10n_br_nfe: Melhorias na transmissão da NF-e (v2)

antoniospneto opened the pull request:
Continuação dos trabalhos feitos nas PRs #2945 e #3065.

As duas PRs são bastante conflitantes, por isso optei por juntar o trabalho aqui.

Resumo das melhorias propostas nesta PR:

- Adição do suporte para transmissão síncrona.
- Melhoria no processo de envio assíncrono para evitar perda de dados em caso de falha.
- Implementação do wizard de consulta de situação para a NF-e.
- Suporte para transmissão normal, utilizando as Sefaz autorizadoras de contingência.
- Outras pequenas correções de usabilidade.

Depende de:

- [x] https://github.com/erpbrasil/erpbrasil.edoc/pull/77
- [x] https://github.com/akretion/nfelib/pull/103

Falta fazer:
- [x] Opção de habilitar a sefaz de contingencia nos outros serviços: Inutilização, Cancelamento, Carta de Correção.
- [x] Implementar mais testes

rvalyi commented:
fiz o merge de https://github.com/akretion/nfelib/pull/103
e fiz o push da nfelib-2.0.6 no pypi

Não sei se isso permite aos testes de rodar, vou mandar rodar...

antoniospneto committed:
[REF] l10n_br_fiscal: fix typos; improve wording

antoniospneto committed:
[IMP] l10n_br_fiscal: adds support for sync communcation

antoniospneto committed:
[IMP] l10n_br_nfe: Various enhancements and fixes

- Added support for contingency mode (SVC)
- Added support for synchronous transmission
- Added NFe status check functionality
- Improved transmission process to ensure information received from SEFAZ is not lost on failure
- Various performance optimizations and bug fixes

nfe

antoniospneto committed:
[REF] l10n_br_account_nfe: rename nfce contingency function

rvalyi commented on l10n_br_nfe/tests/test_nfe_webservices.py:
```diff
@@ -1,20 +1,25 @@
 # Copyright (C) 2023 - TODAY Raphaël Valyi - Akretion
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
 
```
o arquivo tem 165 linhas e vc ta mexendo em 74 linhas, é uma proporçaõ grande do arquivo, é legitimo vc adicionar seu copyright se quiser.

antoniospneto replied:
opa vou por, valeu!

rvalyi replied:
Eu comecei a olhar o seu PR. Eu olhei tudo menos l10n_br_nfe/models/document.py ainda (que é onde tem mais coisas). Nesses outros arquivos me parece OK. Vou revisar o document.py essas dias assim que eu puder...

antoniospneto replied:
Copyright adicionado, valeu @rvalyi

rvalyi reviewed (approved):
LGTM.

@mileo vcs que sao co-autores do modulo e fizeram a maior parte do API de transmissão, alguma posiçao sobre esse refator importante?

cc @renatonlima @marcelsavegnago

marcelsavegnago reviewed (approved):
LGTM

rvalyi commented:
ping @mileo nehnuma objeçao com este PR?

renatonlima commented:
@antoniospneto,

Muito boa as contribuições, há um tempo atrás eu fiz um protótipo para refatorar todo o código do l10n_br_fiscal.document, l10n_br_fiscal.document.eletronic, l10n_br_fiscal.document.workflow, l10n_br_fiscal.event no módulo l10n_br_fiscal, eu acredito que dá para limpar bastante código e estruturar melhor os objetos e métodos porque há muito código no l10n_br_fiscal e deveria estar em outros módulos como l10n_br_nfe, l10n_br_cte, l10n_br_nfse.

Eu diria que seria ótimo a gente fazer esse PR e depois portar para a 15.0 e 16.0 com isso eu iria conseguir integrar com o trabalho de refatoramento que estou fazendo.

rvalyi commented:
@renatonlima @antoniospneto @marcelsavegnago depois deste merge e antes de adicionar mais funcionalidades na gestao dos eventos seria interessante dar uma limpa e extrair este modulo l10n_br_fiscal_edi em fez de aumentar a obesidade morbida do modulo l10n_br_fiscal porem https://github.com/OCA/l10n-brazil/pull/3012 Afinal de conta essa extracao parece bem tranquila.

mileo commented:
> @antoniospneto,
> 
> Muito boa as contribuições, há um tempo atrás eu fiz um protótipo para refatorar todo o código do l10n_br_fiscal.document, l10n_br_fiscal.document.eletronic, l10n_br_fiscal.document.workflow, l10n_br_fiscal.event no módulo l10n_br_fiscal, eu acredito que dá para limpar bastante código e estruturar melhor os objetos e métodos porque há muito código no l10n_br_fiscal e deveria estar em outros módulos como l10n_br_nfe, l10n_br_cte, l10n_br_nfse.
> 
> Eu diria que seria ótimo a gente fazer esse PR e depois portar para a 15.0 e 16.0 com isso eu iria conseguir integrar com o trabalho de refatoramento que estou fazendo.

Vou dar uma testada com NFC-e ainda essa semana e dou um retorno.

rvalyi commented:
> > @antoniospneto,
> > Muito boa as contribuições, há um tempo atrás eu fiz um protótipo para refatorar todo o código do l10n_br_fiscal.document, l10n_br_fiscal.document.eletronic, l10n_br_fiscal.document.workflow, l10n_br_fiscal.event no módulo l10n_br_fiscal, eu acredito que dá para limpar bastante código e estruturar melhor os objetos e métodos porque há muito código no l10n_br_fiscal e deveria estar em outros módulos como l10n_br_nfe, l10n_br_cte, l10n_br_nfse.
> > Eu diria que seria ótimo a gente fazer esse PR e depois portar para a 15.0 e 16.0 com isso eu iria conseguir integrar com o trabalho de refatoramento que estou fazendo.
> 
> Vou dar uma testada com NFC-e ainda essa semana e dou um retorno.

Na real, mesmo que tem coisa de NFCe no diff ele nao alterou nehnuma logica da NFCe, mas ainda assim como eh um refator grande e que mexe nas partes que voces fizeram eh interessante vc confirmar.

rvalyi commented:
@mileo conseguiu ver se tinha algum problema para vcs neste PR?

rvalyi commented on l10n_br_nfe/models/document.py:
```diff
@@ -1023,57 +1064,212 @@ def _generate_key(self):
+                "NF-e data not found when trying to assemble the "
+                "xml with the authorization protocol (nfeProc)"
+            )
+            return None
```
perdi varias horas na migraçao do l10n_br_nfe para a v16 por conta desse if.xml_error_message return de antes (por causa da falta traducao do pais Brazil que dava um erro). Sera se nao seria possivel fazer um _logger.info dos xml_error_message se tiver? (a nao ser que teria algo melhor para se fazer tipo como algum tipo de ValidationError ou gestao de excecao sem tb impedir de escrever o que tem que escrever no banco, mas um log me parece o minimo).

antoniospneto replied:
Opa @rvalyi  eu mudei essa parte do código para a linha 1187, mas ele ainda existe, eu vou acrescentar um logger ali então com a mensagem "Envio abortado pois existem erros de validação no XML do documento" e na mensagem eu imprimo o conteudo do xml_error_message
Isso ?

rvalyi replied:
alguma coisa assim por favor (logando em ingles e com camada de traducao). Nao tenho certeza como esta o comportamento pela interface (talvez esta OK), mas na hora de trabalhar nos testes esse retorno silencioso era foda para descobrir depois.

rvalyi replied:
ja que esse ponto nem melhorou nem piorou no PR, eu vou voltar a aprovar o PR, pois eu acho que o PR eh bem importante e ja representa um ganho muito grande.

mileo commented:
> @mileo conseguiu ver se tinha algum problema para vcs neste PR?

Estamos validando com um cliente esse pr hj.

rvalyi commented:
@mileo pedi seu retorno porque são vcs que fizeram esse codigo da transmissão da NFe sendo refatorado aqui. Mas tb vejo que todos estamos achando o refator top, vc pretende testar ainda ou podemos fazer o merge?

Veja que uma coisa é botar coisas novas como recuperação de crédito, outra coisa é botar mais qualidade no codigo legacy que ainda temos... Essa segunda coisa tem que ser muito mais prioritária antes de espelhar código mais ou menos em branches diferentes onde cada um vai depois querer refatorar de um jeito diferente para atender X ou Y...

mileo commented:
> @mileo pedi seu retorno porque são vcs que fizeram esse codigo da transmissão da NFe sendo refatorado aqui. Mas tb vejo que todos estamos achando o refator top, vc pretende testar ainda ou podemos fazer o merge?
> 
> Veja que uma coisa é botar coisas novas como recuperação de crédito, outra coisa é botar mais qualidade no codigo legacy que ainda temos... Essa segunda coisa tem que ser muito mais prioritária antes de espelhar código mais ou menos em branches diferentes onde cada um vai depois querer refatorar de um jeito diferente para atender X ou Y...

Boa noite, ainda estamos revisando, tem uma pessoa do time validando o problema da NFC-e e estamos testando esse PR em um cliente desde a semana passada, como tem bastante código e depois de mergado fica complicado trabalhar sem ele.

mileo commented:
/ocabot merge major
